src/components/FileInput/index.tsx

```diff
@@ -0,0 +1,215 @@
+import React, { useState, useCallback } from 'react';
+import { _cs } from '@togglecorp/fujs';
+import { MdFileUpload } from 'react-icons/md';
+
+import useUiModeClassName from '../../hooks/useUiModeClassName';
+import InputContainer, { Props as InputContainerProps } from '../InputContainer';
+import RawInput, { Props as RawInputProps } from '../RawInput';
+import styles from './styles.css';
+
+type InheritedProps<T> = (Omit<InputContainerProps, 'input'> & Omit<RawInputProps<T>, 'onChange'>);
+export interface Props<T extends string> extends InheritedProps<T> {
+    inputElementRef?: React.RefObject<HTMLInputElement>;
+    inputClassName?: string;
+    showStatus?: boolean;
+    onChange?: (files: File[]) => void;
+}
+
+function FileInput<T extends string>(props: Props<T>) {
+    const {
+        actions,
+        actionsContainerClassName,
+        className,
+        disabled,
+        error,
+        errorContainerClassName,
+        hint,
+        hintContainerClassName,
+        icons,
+        iconsContainerClassName,
+        inputSectionClassName,
+        inputContainerClassName,
+        label,
+        labelContainerClassName,
+        readOnly,
+        uiMode,
+        inputElementRef,
+        containerRef,
+        inputSectionRef,
+        inputClassName,
+        value,
+        onChange,
+        showStatus = true,
+        name,
+        multiple,
+        accept,
+        ...fileInputProps
+    } = props;
+
+    const uiModeClassName = useUiModeClassName(uiMode, styles.light, styles.dark);
+
+    const dragEnterRef = React.useRef(0);
+    const [status, setStatus] = useState<string>('No file chosen');
+    const [isBeingDraggedOver, setIsBeingDraggedOver] = React.useState(false);
+
+    const isValidFile = useCallback((fileName: string, mimeType: string, acceptString?: string) => {
+        // if there is no accept string, anything is valid
+        if (!acceptString) {
+            return true;
+        }
+        const extensionMatch = /\.\w+$/.exec(fileName);
+        const mimeMatch = /^.+\//.exec(mimeType);
+
+        const fileTypeList = acceptString.split(/,\s+/);
+        return fileTypeList.some((fileType) => {
+            // check mimeType such as image/png or image/*
+            if (mimeType === fileType || (!!mimeMatch && `${mimeMatch[0]}*` === fileType)) {
+                return true;
+            }
+            return !!extensionMatch && extensionMatch[0].toLowerCase() === fileType.toLowerCase();
+        });
+    }, []);
+
+    const handleDragEnter = useCallback(() => {
+        if (dragEnterRef.current === 0) {
+            setIsBeingDraggedOver(true);
+        }
+
+        dragEnterRef.current += 1;
+    }, [setIsBeingDraggedOver]);
+
+    const handleDragLeave = useCallback(() => {
+        dragEnterRef.current -= 1;
+
+        if (dragEnterRef.current === 0) {
+            setIsBeingDraggedOver(false);
+        }
+    }, [setIsBeingDraggedOver]);
```

Comment:
The drag handling should be reused from DropContainer.
Maybe we need to create a useDrop hook

Reply:
The DropContainer only works for text.

Reply:
But we can create a useDrop hook that's more generic that can be used on both cases.